react/src/Login.tsx

```diff
@@ -56,53 +138,132 @@ export default function LoginForm({
+                    className={classes.textField}
+                    label="Username"
+                    onChange={bind(setUsername)}
+                />
+                <TextField
+                    required
+                    variant="filled"
+                    fullWidth
+                    margin="normal"
+                    className={classes.textField}
+                    type="password"
+                    label="Password"
+                    onChange={bind(setPassword)}
+                    autoComplete="current-password"
+                />
+                <Button
+                    variant="contained"
+                    color="primary"
+                    className={classes.button}
+                    type="submit"
+                    onClick={authenticate}
+                >
+                    Sign in
+                </Button>
+            </form>
+        </>
+    );
+}
+
+/**
+ * A button whose href is derived from the backend API endpoint.
+ * Used for redirecting to OAuth login.
+ */
+function OauthLoginForm() {
+    const classes = useStyles();
+    const [loginUrl, setLoginUrl] = useState("");
+
     useEffect(() => {
-        document.title = `Sign in | ${process.env.REACT_APP_NAME}`;
+        if (process.env.REACT_APP_API_ENDPOINT) {
+            // Build login link with redirect route
+            const redirect = new URL(`${window.location.origin}/oidc_callback`);
+            const login = new URL(`${process.env.REACT_APP_API_ENDPOINT}/api/login`);
+            login.searchParams.append("redirect_uri", redirect.href);
+            setLoginUrl(login.href);
+        }
     }, []);
+
+    return (
+        <>
+            <Typography variant="h5" component="h2" gutterBottom className={classes.center}>
+                Sign in to {process.env.REACT_APP_NAME}
+            </Typography>
+            <form>
+                <Button
+                    variant="contained"
+                    color="primary"
+                    className={classes.button}
+                    type="submit"
+                    disabled={!loginUrl}
+                    href={loginUrl}
+                    fullWidth
+                >
+                    Sign in using OAuth
+                </Button>
+            </form>
+        </>
+    );
+}
+
+/**
+ * Wrapper component for rendering all possible login forms,
+ * based on whether OAuth / OIDC is enabled or disabled.
+ */
+export default function LoginPage(props: LoginProps) {
     const classes = useStyles();
+
+    useEffect(() => {
+        document.title = `Sign in | ${process.env.REACT_APP_NAME}`;
+    }, []);
+
+    const OAuthCallback = useMemo(
+        () => (props.oauth ? <OIDCRedirectHandler {...props} /> : <Redirect to="/" />),
+        [props]
+    );
+
+    const FormDisplay: JSX.Element = useMemo(() => {
+        if (props.oauth === true) {
+            return <OauthLoginForm />;
+        }
+        return <LoginForm {...props} />;
+    }, [props]);
```

Comment:
I see, this isn't a proper functional component because it's actually useMemo. Is there a reason for using that over just the body of this as a component (or inlined below?)

Reply:
I figured memoizing it made sense since `props.oauth` should never change for the duration that React is being served. I can move this to inline instead.